api/apps/v1alpha1/nemo_entitystore_types.go

```
@@ -0,0 +1,758 @@
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []NemoEntitystore `json:"items"`
+}
+
+// GetPVCName returns the name to be used for the PVC based on the custom spec
+// Prefers pvc.Name if explicitly set by the user in the NemoEntitystore instance
+func (n *NemoEntitystore) GetPVCName(pvc PersistentVolumeClaim) string {
+	pvcName := fmt.Sprintf("%s-pvc", n.GetName())
+	if pvc.Name != "" {
+		pvcName = pvc.Name
+	}
+	return pvcName
+}
+
+// GetStandardSelectorLabels returns the standard selector labels for the NemoEntitystore deployment
+func (n *NemoEntitystore) GetStandardSelectorLabels() map[string]string {
+	return map[string]string{
+		"app":                        n.Name,
+		"app.kubernetes.io/name":     n.Name,
+		"app.kubernetes.io/instance": n.Name,
+	}
+}
+
+// GetStandardLabels returns the standard set of labels for NemoEntitystore resources
+func (n *NemoEntitystore) GetStandardLabels() map[string]string {
+	return map[string]string{
+		"app.kubernetes.io/name":             n.Name,
+		"app.kubernetes.io/instance":         n.Name,
+		"app.kubernetes.io/operator-version": os.Getenv("OPERATOR_VERSION"),
+		"app.kubernetes.io/part-of":          "nemo-entitystore-service",
+		"app.kubernetes.io/managed-by":       "k8s-nim-operator",
+	}
+}
+
+// GetStandardEnv returns the standard set of env variables for the NemoEntitystore container
+func (n *NemoEntitystore) GetStandardEnv() []corev1.EnvVar {
+	passwordKey := "password"
+	if n.Spec.EntityStoreParams.DBSecret == "" {
+		passwordKey = n.Spec.EntityStoreParams.DBSecretPasswordKey
+	}
+
+	// add standard env required for NIM service
+	envVars := []corev1.EnvVar{
+		{
+			Name:  "APP_VERSION",
+			Value: n.Spec.EntityStoreParams.AppVersion,
+		},
+		{
+			Name: "POSTGRES_PASSWORD",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					Key: passwordKey,
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: n.Spec.EntityStoreParams.DBSecret,
```

We should probably throw an error in the reconciler if this field is not set.

---

I wanted to start a separate thread on this. We need a a proper validation-webhook to validate these CRs instead of allowing them to be created empty

---

yes, validation webhooks are ideal but mostly used for complex validations (like app version dependent checks and compatibility etc). They also provide better user experience as they fail fast on invalid params. However k8s requires webhooks to use https, hence we need to either depend on cert-manager to be deployed or manage self-signed certs manually. In the meantime, we should add validations in the controller and set proper status/conditions to indicate these error.

---

I've reworked the validations in a new commit.